build.sbt

```
@@ -21,8 +26,54 @@ val rewrites = (project in file("rewrites"))
     }.dependsOn(Compile / compile).value
   )
 
+val incompatSettings = inConfig(CompileBackward)(Defaults.compileSettings) ++ 
+  Seq(
+    scalaVersion := dotty,
+    crossScalaVersions := List(scala213, dotty),
+    Compile / unmanagedSourceDirectories := Seq(baseDirectory.value / s"src/main/scala"),
+    CompileBackward / unmanagedSourceDirectories := Seq(baseDirectory.value / s"src/main/scala-2.13"),
+    CompileBackward / managedClasspath := (managedClasspath in Compile).value,
+    Test / test := Def.task { 
+      checkIncompatibility(
+        name.value,
+        scalaVersion.value,
+        (CompileBackward / compile).result.value,
+        state.value.log
+      )
+    }.dependsOn(Compile / compile).value
+  )
+
+lazy val incompat = (project in file("incompat"))
+  .configs(CompileBackward)
+  .aggregate(typeInfer1)
+
+lazy val typeInfer1 = (project in file("incompat/type-infer-1"))
+  .settings(incompatSettings)
+
 def copySources(inputDir: File, outputDir: File): Seq[File] = {
   if (outputDir.exists) FileUtils.deleteDirectory(outputDir)
   FileUtils.copyDirectory(inputDir, outputDir)
   outputDir.listFiles.filter(_.isFile)
 }
+
+def checkIncompatibility(name: String, scalaVersion: String, compileResult: Result[CompileAnalysis], log: Logger): Unit = {
+  CrossVersion.partialVersion(scalaVersion).foreach {
+    case (0, _) => compileResult match {
+      case Value(_) => 
+        log.error("Compilation has succeeded but failure was expected.")
+        log.error(s"The '$name' incompatibility is probably fixed, in version $scalaVersion.")
+        throw new IllegalStateException()
```

The recommended way to report an intentional error in an sbt task is to throw a `MessageOnlyException` with the message as a parameter:
```suggestion
        throw new MessageOnlyException(
          "Compilation has succeeded but failure was expected. " +
          s"The '$name' incompatibility is probably fixed, in version $scalaVersion."
        )
```